devito/arguments.py

```diff
@@ -158,198 +107,350 @@ class PtrArgument(Argument):
+            
+        # Record dependencies in Dimensions
+        dimension_parameter_mapper = {}
+        for dim, deps in dimension_dependency_mapper.items():
+            dimension_parameter_mapper[dim] = DimensionParameter(dim, deps)
+
+        # Dimensions that are in parameters but not directly referenced in the expressions
+        for dim in [x for x in parameters if isinstance(x, Dimension) and x not in dimension_dependency_mapper.keys()]:
+            dimension_parameter_mapper[dim] = DimensionParameter(dim, [])
+
+        for dim in [x for x in parameters if isinstance(x, Dimension) and x.is_Stepping]:
+            dimension_parameter_mapper[dim.parent].dependencies.append(Dependency(Dependency.GETS_VALUE_FROM, dimension_parameter_mapper[dim]))
+
+        dimension_parameters = list(dimension_parameter_mapper.values())
+        
+        # Pass Dimensions
+        scalar_arguments = []
+        for dimension_parameter in dimension_parameters:
+            scalar_arguments += ArgumentVisitor().visit(dimension_parameter)
+
+        other_arguments = [ArgumentVisitor().visit(x) for x in parameters if x not in tensor_arguments + scalar_arguments]
+            
+        return tensor_arguments + dimension_parameters + scalar_arguments + other_arguments
+
+    def _extract_children_of_composites(self, kwargs):
+        new_params = {}
+        # If we've been passed CompositeFunction objects as kwargs,
+        # they might have children that need to be substituted as well.
+        for k, v in kwargs.items():
+            if isinstance(v, CompositeFunction):
+                orig_param_l = [i for i in self.parameters if i.name == k]
+                # If I have been passed a parameter, I must have seen it before
+                if len(orig_param_l) == 0:
+                    raise InvalidArgument("Parameter %s does not exist in expressions " +
+                                          "passed to this Operator" % k)
+                # We've made sure the list isn't empty. Names should be unique so it
+                # should have exactly one entry
+                assert(len(orig_param_l) == 1)
+                orig_param = orig_param_l[0]
+                # Pull out the children and add them to kwargs
+                for orig_child, new_child in zip(orig_param.children, v.children):
+                    new_params[orig_child.name] = new_child
+        kwargs.update(new_params)
+        return kwargs
+
+    def _dle_arguments(self, dim_sizes):
+        # Add user-provided block sizes, if any
+        dle_arguments = OrderedDict()
+        autotune = True
+        for i in self.dle_arguments:
+            dim_size = dim_sizes.get(i.original_dim.name, None)
+            if dim_size is None:
+                error('Unable to derive size of dimension %s from defaults. '
+                      'Please provide an explicit value.' % i.original_dim.name)
+                raise InvalidArgument('Unknown dimension size')
```

Comment:
Question: this is something I was seeing in ``function.py`` too. On (user) errors, we have both ``error`` and ``raise``. Why do we need both? I think you can just stick the error message within ``raise``

Reply:
This is how the code in master currently is. I just moved it from `operator.py` to `arguments.py`